### PR TITLE
cmake: Fix dependencies so builds are sparse again

### PR DESCRIPTION
**Pull request by mped-oticon**
Rebuilding unchanged Zephyr applications using GNU ld, has the last 4 months performed linking again.
I.e. running cmake && ninja over an unchanged application has cost linking time.
This PR fixes that, i.e. makes the build sparse again so linker is not invoked again needlessly.
ld/target.cmake had drifted behind other linkers, lacking ${base_name}.

GCC can build dependency files which ninja picks up.
GCC is told to generate a dep file that depends on a target that is
never created: The file is created but under the basename zephyr folder.
Thus the dependency check always failed.

This also realigns ld with lld and arcmwdt:

```
$ rg -tcmake base_name
cmake/linker/arcmwdt/target.cmake
33:  get_filename_component(base_name ${CMAKE_CURRENT_BINARY_DIR} NAME)
49:    -MD -MF ${linker_script_gen}.dep -MT ${base_name}/${linker_script_gen}

cmake/linker/ld/target.cmake
67:    get_filename_component(base_name ${CMAKE_CURRENT_BINARY_DIR} NAME)
81:      -MD -MF ${linker_script_gen}.dep -MT ${base_name}/${linker_script_gen}

cmake/linker/lld/target.cmake
32:  get_filename_component(base_name ${CMAKE_CURRENT_BINARY_DIR} NAME)
45:    -MD -MF ${linker_script_gen}.dep -MT ${base_name}/${linker_script_gen}
```

Finally, use copy rather than rename as renaming can break dependencies. Copying is more idempotent.

### CODE REVIEW

File: cmake/linker/ld/target.cmake [hunk 1]
```diff
@@ -77,7 +78,7 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
       COMMAND ${CMAKE_C_COMPILER}
       -x assembler-with-cpp
       ${NOSYSDEF_CFLAG}
-      -MD -MF ${linker_script_gen}.dep -MT ${linker_script_gen}
+      -MD -MF ${linker_script_gen}.dep -MT ${base_name}/${linker_script_gen}
```
Comment (tejlmand) [inline cmake/linker/ld/target.cmake:81]: This was removed here: https://github.com/zephyrproject-rtos/zephyr/commit/a0ba67df63577a91ccd00d6a8c229b74928216e3

Are you perhaps having `CMP0116` set to OLD behavior in your system ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
With this PR, I get:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/hello_world
Loading Zephyr default modules (Zephyr repository).
-- Application: /projects/github/ncs/zephyr/samples/hello_world
...
-- Build files have been written to: /projects/github/ncs/zephyr/build
$ ninja -Cbuild
ninja: Entering directory `build'
[1/158] Preparing syscall dependency handling

[2/158] Generating include/generated/version.h
-- Zephyr version: 3.0.99 (/projects/github/ncs/zephyr), build: zephyr-v3.0.0-1040-g04faaeeb975e
[148/158] Linking C executable zephyr/zephyr_pre0.elf

[152/158] Linking C executable zephyr/zephyr_pre1.elf

[158/158] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19288 B         1 MB      1.84%
            SRAM:        5504 B       256 KB      2.10%
        IDT_LIST:          0 GB         2 KB      0.00%
$ ninja -Cbuild
ninja: Entering directory `build'
[3/11] Linking C executable zephyr/zephyr_pre0.elf

[6/11] Linking C executable zephyr/zephyr_pre1.elf

[11/11] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19288 B         1 MB      1.84%
            SRAM:        5504 B       256 KB      2.10%
        IDT_LIST:          0 GB         2 KB      0.00%

```
instead of (zephyr/main):
```
$ ninja -Cbuild
ninja: Entering directory `build'
[1/158] Preparing syscall dependency handling

[2/158] Generating include/generated/version.h
-- Zephyr version: 3.0.99 (/projects/github/ncs/zephyr), build: zephyr-v3.0.0-1095-g1c7c4506d85c
[148/158] Linking C executable zephyr/zephyr_pre0.elf

[152/158] Linking C executable zephyr/zephyr_pre1.elf

[158/158] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19288 B         1 MB      1.84%
            SRAM:        5504 B       256 KB      2.10%
        IDT_LIST:          0 GB         2 KB      0.00%
$ ninja -Cbuild
ninja: Entering directory `build'
ninja: no work to do.
$
```

Could you check the `CMP0116` setting in your system ?

Zephyr removed `CMP0116 OLD` here: https://github.com/zephyrproject-rtos/zephyr/pull/41229

**Comment by mped-oticon** [comment]
Thanks @tejlmand. PR updated as instructed.
(We were stuck on an older cmake which explains the difference)

**Review by tejlmand — APPROVED** [review]
lgtm

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**